system/jlib/jptree.cpp

```diff
@@ -8540,12 +8555,128 @@ static void holdLoop()
+
+                /* swapin before callbacks called, but before releasing crit.
+                 * That way the CB can see the old/diffs and act on them, but any
+                 * code calling e.g. getComponentConfig() will see new.
+                 */
+                componentConfiguration.setown(std::get<1>(result));
+                globalConfiguration.setown(std::get<2>(result));
+
+                /* NB: we are still holding 'configCS' at this point, blocking all other thread access.
+                   However code in callbacks may call e.g. getComponentConfig() and re-enter the crit */
+                for (const auto &item: notifyConfigUpdates)
+                    item.second(oldComponentConfiguration, oldGlobalConfiguration);
+
+                absoluteConfigFilename.set(std::get<0>(result).c_str());
+            }
+        };
+
+        fileWatcher.setown(createFileEventWatcher(updateFunc));
+
+        // watch the path, not the filename, because the filename might not be seen if directories are moved, softlinks are changed..
+        StringBuffer path, filename;
+        splitFilename(absoluteConfigFilename, nullptr, &path, &filename, &filename);
+        configFilename.set(filename);
+        fileWatcher->add(path, FileWatchEvents::anyChange);
+        fileWatcher->start();
+    }
+    unsigned addNotifyFunc(ConfigUpdateFunc notifyFunc)
+    {
+        CriticalBlock b(notifyFuncCS);
+        notifyFuncId++;
+        notifyConfigUpdates[notifyFuncId] = notifyFunc;
+        return notifyFuncId;
+    }
+    bool removeNotifyFunc(unsigned funcId)
+    {
+        CriticalBlock b(notifyFuncCS);
+        return notifyConfigUpdates.erase(funcId) > 0;
+    }
+};
+
+static Owned<CConfigUpdater> configFileUpdater;
+
+unsigned installConfigUpdateHook(ConfigUpdateFunc notifyFunc)
+{
+#ifdef _CONTAINERIZED
+    if (!configFileUpdater)
+        WARNLOG("installConfigUpdateHook(): configuration updater not installed");
+    else
+        return configFileUpdater->addNotifyFunc(notifyFunc);
+#endif
+    return 0;
+}
+
+void removeConfigUpdateHook(unsigned notifyFuncId)
+{
+#ifdef _CONTAINERIZED
+    if (!configFileUpdater)
+    {
+        WARNLOG("removeConfigUpdateHook(): configuration updater not installed");
+        return;
+    }
+    if (!configFileUpdater->removeNotifyFunc(notifyFuncId))
+        WARNLOG("removeConfigUpdateHook(): notifyFuncId %u not installed");
+#endif
+}
+#else
+unsigned installConfigUpdateHook(std::function<void ()> notifyFunc)
+{
+    return 0;
+}
+
+void removeConfigUpdateHook(unsigned notifyFuncId)
+{
+}
+#endif
+
+
+static std::tuple<std::string, IPropertyTree *, IPropertyTree *> doLoadConfiguration(IPropertyTree *componentDefault, const char * * argv, const char * componentTag, const char * envPrefix, const char * legacyFilename, IPropertyTree * (mapper)(IPropertyTree *), const char *altNameAttribute)
+{
+    Linked<IPropertyTree> newComponentConfig = createPTreeFromIPT(componentDefault);
```

Comment:
I think this should be Owned<>

Reply:
@ghalliday  fixed, but forgot to tag you.